presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java

```diff
@@ -0,0 +1,202 @@
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.planner.plan.TableScanNode;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+import static com.google.common.base.Verify.verify;
+
+public class DynamicFilter
+{
+    private static final Logger log = Logger.get(DynamicFilter.class);
+
+    private final List<Descriptor> descriptors;
+    private final TableScanNode tableScanNode;
+    private final SettableFuture<TupleDomain<ColumnHandle>> resultFuture;
+
+    private final Consumer<TupleDomain<ColumnHandle>> tupleDomainConsumer;
+    private TupleDomain<ColumnHandle>[] partitions;
+    private AtomicInteger partitionCounter;
+
+    public DynamicFilter(List<Descriptor> descriptors, TableScanNode tableScanNode, int partitionCount)
+    {
+        this.descriptors = descriptors;
+        this.tableScanNode = tableScanNode;
+        this.resultFuture = SettableFuture.create();
+
+        this.partitions = new TupleDomain[partitionCount];
+        this.partitionCounter = new AtomicInteger();
+        this.tupleDomainConsumer = tupleDomain -> {
+            if (partitions == null) {
+                return;
+            }
+            if (tupleDomain.isAll()) {
+                verify(resultFuture.set(TupleDomain.all()), "dynamic filter result is provided more than once");
+                partitions = null;  // drop all collected predicates if one of partitions' results is too large.
+                return;
+            }
+            // Called by each DynamicFilterCollectOperator instance (when collection is over).
+            final int partitionIndex = this.partitionCounter.getAndIncrement();
+            verify(partitionIndex < partitions.length);
+            partitions[partitionIndex] = tupleDomain;
+            if (partitionIndex == partitions.length - 1) {
+                TupleDomain<ColumnHandle> result = Arrays.stream(partitions).reduce(TupleDomain.none(), TupleDomain::columnWiseUnion);
+                // NOTE: may result in a bit more relaxed constraint if there are multiple columns and multiple rows.
+                // See the comment at TupleDomain::columnWiseUnion() for more details.
+                verify(resultFuture.set(result), "dynamic filter result is provided more than once");
+            }
+        };
+    }
+
+    public static Optional<DynamicFilter> create(JoinNode planNode, TypeProvider types, int partitionCount)
+    {
+        PlanNode currentNode = planNode;
+        PlanNode buildNode = planNode.getRight();
+
+        Map<String, Symbol> filterSymbols = new HashMap<>(); // mapping filter ID to its probe-side symbol
+        while (currentNode != null) {
```

Comment:
you can use `io.prestosql.sql.planner.optimizations.PlanNodeSearcher` to search for all table scans with filter on top (which have matching dynamic filters)

Reply:
https://github.com/prestosql/presto/commit/4561c62d7d5b26a2fa12c623d4f279daac7360ea#diff-d3d099fe2c1d2a89408b9234c0d7ab29R127